### PR TITLE
Add Observer to EligibilityService

### PR DESCRIPTION
**Pull request by moritzhaller**
Resolves https://github.com/brave/brave-browser/issues/21666

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_federated/eligibility_service.cc [hunk 1]
```diff
@@ -20,22 +20,53 @@ EligibilityService::~EligibilityService() {
...
+  if (is_eligible_ == IsEligibile()) {
+    return;
+  }
+
+  is_eligible_ = !is_eligible_;
+  NotifyObservers(is_eligible_);
 }
 
-bool EligibilityService::IsConnectionWifiOrEthernet() {
+bool EligibilityService::IsConnectedToWifiOrEthernet() {
```
Comment (tmancey) [inline components/brave_federated/eligibility_service.cc:56]: `const`

File: components/brave_federated/eligibility_service.cc [hunk 2]
```diff
@@ -20,22 +20,53 @@ EligibilityService::~EligibilityService() {
...
+  observers_.RemoveObserver(observer);
+}
+
+void EligibilityService::NotifyObservers(bool is_eligible) {
+  for (auto& observer : observers_) {
+    observer.OnEligibilityChanged(is_eligible);
+  }
+}
+
+bool EligibilityService::IsEligibile() {
```
Comment (tmancey) [inline components/brave_federated/eligibility_service.cc:41]: `const`

File: components/brave_federated/eligibility_service_observer.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_OBSERVER_H_
+#define BRAVE_COMPONENTS_BRAVE_FEDERATED_ELIGIBILITY_SERVICE_OBSERVER_H_
+
+#include "base/observer_list.h"
```
Comment (tmancey) [inline components/brave_federated/eligibility_service_observer.h:9]: should be `#include "base/observer_list_types.h"`

**Review by tmancey — APPROVED** [review]
LGTM++

**Review by aseren — APPROVED** [review]

**Review by LorenzoMinto — APPROVED** [review]

**Pull request merged**